source/base/kokkos.cc

```diff
@@ -0,0 +1,61 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2022 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/kokkos.h>
+
+#include <deal.II/lac/la_parallel_vector.h>
+#include <deal.II/lac/vector_memory.h>
+
+#include <Kokkos_Core.hpp>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Impl
+{
+  void
+  ensure_kokkos_initialized()
+  {
+    if (!Kokkos::is_initialized())
+      {
+        GrowingVectorMemory<
+          LinearAlgebra::distributed::Vector<double, MemorySpace::Host>>{};
+        GrowingVectorMemory<
+          LinearAlgebra::distributed::Vector<float, MemorySpace::Host>>{};
+        GrowingVectorMemory<
+          LinearAlgebra::distributed::Vector<double, MemorySpace::Device>>{};
+        GrowingVectorMemory<
+          LinearAlgebra::distributed::Vector<float, MemorySpace::Device>>{};
+        Kokkos::push_finalize_hook(
+          GrowingVectorMemory<
+            LinearAlgebra::distributed::Vector<double, MemorySpace::Host>>::
+            release_unused_memory);
+        Kokkos::push_finalize_hook(
+          GrowingVectorMemory<
+            LinearAlgebra::distributed::Vector<float, MemorySpace::Host>>::
+            release_unused_memory);
+        Kokkos::push_finalize_hook(
+          GrowingVectorMemory<
+            LinearAlgebra::distributed::Vector<double, MemorySpace::Device>>::
+            release_unused_memory);
+        Kokkos::push_finalize_hook(
+          GrowingVectorMemory<
+            LinearAlgebra::distributed::Vector<float, MemorySpace::Device>>::
+            release_unused_memory);
+        Kokkos::initialize();
+        std::atexit(Kokkos::finalize);
```

Comment:
Could this be problematic if one of our dependencies uses kokkos and also calls the init/finalize functions?

Reply:
Whoever initializes `Kokkos` should also finalize it. If another dependency or user code initializes `Kokkos` before, then they are also responsible for cleaning up and this function does nothing. If this function initializes `Kokkos` no other dependency can initialize `Kokkos` and shouldn't finalize it.

In the end, it's always preferable if user code initializes `Kokkos` but we can't really force that easily (without breaking every code built on top of `deal.II`).